docs/design-proposals/apiserver/Data-Partition-Design.md

```diff
@@ -0,0 +1,198 @@
+---
+title: API Server Data Partition Design for CloudFabric
+authors:
+  - "@sindica"
+---
+
+# API Server Data Partition Design for CloudFabric
+
+## Table of Contents
+
+## Motivation
+
+Community version of Kubernetes can take up 5K nodes per cluster, each 
+nodes has 30 pods without significant performance issue. One of the goals 
+of CloudFabric is to greatly increase the number of nodes that can be 
+managed by a single cluster.
+
+With current community design of each api server get all data from 
+backend storage (ETCD for now), api server becomes one of the bottle 
+necks in CloudFabric scaling up.
+
+This document will address how to partition data across api server instances.
+
+### Goals
+
+* Primary: Define a mechanism that data from backend storage can be effectively
+partitioned across multiple api server instances
+* Primary: Allows clients to get all data continuously without service interruption
+* Primary: Allows users to manage data partition without service interruption
+* Secondary: Single parition cluster performance shall aligned with community version
+* Secondary: System will automatically adjust data partition based on data volume and 
+traffic pattern
+
+### Non-Goals
+
+* Data partition of backend storage is not covered in this design doc
+* Scaling out of controller framework is not discussed in this doc
+* Automatically adjust data partition configuration based on data and traffic pattern
+will not be addressed in this design doc 
+
+## Proposal
+
+### API Changes
+
+* We added a top level Kubernetes object named '**datapartitionconfigs**' with path
+'**/api/v1/datapartitionconfigs**'.
+ - This allows each api server to get data partition setting assigned to it
+ - This allows admin to view and adjust data partition across api servers manually
+ - This allows managing component to list and adjust data partition across api servers
+based on data volume and traffic automatically
+
+```yaml
+apiVersion: v1
+kind: DataPartitionConfig
+serviceGroupId: "1"
+startTenant: "A"
+isStartTenantValid: false 
+endTenant: "m"
+isEndTenantValid: true
+metadata:
+  name: "partition-1"
+``` 
+
+```text
+$ curl http://127.0.0.1:8080/api/v1/datapartitionconfigs
+{
+  "kind": "DataPartitionConfigList",
+  "apiVersion": "v1",
+  "metadata": {
+    "selfLink": "/api/v1/datapartitionconfigs",
+    "resourceVersion": "670"
+  },
+  "items": [
+    {
+      "metadata": {
+        "name": "partition-1",
+        "selfLink": "/api/v1/datapartitionconfigs/partition-1",
+        "uid": "c28d0dcf-52af-4794-90e8-160c353d22ba",
+        "hashKey": 1479568363250919565,
+        "resourceVersion": "438",
+        "creationTimestamp": "2020-04-04T01:54:12Z"
+      },
+      "startTenant": "A",
+      "endTenant": "m",
+      "isEndTenantValid": true,
+      "serviceGroupId": "1"
+    }
+  ]
+}
+```
+
+* We introduced a field '**ServiceGroupId**' into '**EndpointSubset**'.
```

Comment:
I'm thinking if there is a more lightweight way to do this. But we can discuss this later and include any potential changes in subsequent updates. I will approve this for 330 release.